src/host-config/index.ts

```diff
@@ -0,0 +1,128 @@
+
+export * from './proxy';
+export * from './types';
+
+const hostnamePath = pathOnRoot('/etc/hostname');
+
+export async function readHostname() {
+	const hostnameData = await fs.readFile(hostnamePath, 'utf-8');
+	return hostnameData.trim();
+}
+
+export async function setHostname(val: string) {
+	let hostname = val;
+	// If hostname is an empty string, return first 7 digits of device uuid
+	if (!val) {
+		const uuid = await config.get('uuid');
+		hostname = uuid?.slice(0, 7) as string;
+	}
+	// Changing the hostname on config.json will trigger
+	// the OS config-json service to restart the necessary services
+	// so the change gets reflected on containers
+	await config.set({ hostname });
+}
+
+export function parse(
+	conf: unknown,
+): HostConfiguration | LegacyHostConfiguration | null {
+	const decoded = HostConfiguration.decode(conf);
+
+	if (isRight(decoded)) {
+		return decoded.right;
+	} else {
+		log.warn(
+			['Malformed host config detected, things may not behave as expected:']
+				.concat(Reporter.report(decoded))
+				.join('\n'),
+		);
+		// We haven't strictly validated user input since introducing the API,
+		// endpoint, so accept configs where values may not be of the right type
+		// but have the right shape for backwards compatibility.
+		const legacyDecoded = LegacyHostConfiguration.decode(conf);
+		if (isRight(legacyDecoded)) {
+			return legacyDecoded.right;
+		}
+	}
+	return null;
+}
+
+// Exported for unit tests
+export function patchProxy(
+	currentConf: RedsocksConfig,
+	inputConf: Partial<{
+		redsocks: Partial<ProxyConfig>;
+	}>,
+): RedsocksConfig {
+	const patchedConf: RedsocksConfig = {};
+
+	// If input is empty, redsocks config should be removed
+	if (!inputConf || Object.keys(inputConf).length === 0) {
+		return patchedConf;
+	}
+
+	if (inputConf.redsocks && Object.keys(inputConf.redsocks).length > 0) {
+		// This method assumes that currentConf is a full ProxyConfig
+		// for backwards compatibility, so patchedConf.redsocks should
+		// also be a full ProxyConfig.
+		const patched = {
+			...currentConf.redsocks,
+			...inputConf.redsocks,
+		} as Record<string, any>;
+		patchedConf.redsocks = patched as ProxyConfig;
+	}
+	return patchedConf;
+}
+
+export async function patch(
+	conf: HostConfiguration | LegacyHostConfiguration,
+	force: boolean = false,
+): Promise<void> {
+	const apps = await applicationManager.getCurrentApps();
+	const appIds = Object.keys(apps).map((strId) => parseInt(strId, 10));
+
+	if (conf.network.hostname != null) {
+		await setHostname(conf.network.hostname);
+	}
+
+	if (conf.network.proxy != null) {
+		const targetConf = conf.network.proxy;
+		// It's possible for appIds to be an empty array, but patch shouldn't fail
+		// as it's not dependent on there being any running user applications.
+		return updateLock.lock(appIds, { force }, async () => {
+			const { noProxy: _ignore, ...currentConf } =
+				(await readProxy()) || ({} as any);
```

Comment:
This `||` is not necessary if `readProxy` can return an empty object

Reply:
I was wrong before about `readProxy` being able to return an empty object, but I don't t think we should use `as any` here as it disables all type checking potentially masking bugs. The right way might just be to use an if 

```ts
const proxyConf = await readProxy();
let currentConf: ProxyConfig | undefined = undefined;
if (proxyConf) {
	delete proxyConf.noProxy;
	currentConf = proxyConf;
}
```

Reply:
I applied the suggested change [here](https://github.com/balena-os/balena-supervisor/pull/2329/commits/53f5641ef1cd2e5374f55a50a601fb80d4872511#diff-aaac9f5c9b6707e5fe6448bad4487ef43f9686b75b9728aebc50aab095e0c4f4R104-R109).